DYN-5965: introduce new string from nodes that accept numeric format specifiers

mjkkirschner opened the pull request:
### Purpose

[DYN-5965](https://jira.autodesk.com/browse/DYN-5965)
[DYN-6557](https://jira.autodesk.com/browse/DYN-6557)



this PR does the following:
* adds some new overloads for the string conversion methods that the `String From Object` and `String From Array` nodes call internally.
* modifies the built-in methods and string from nodes to accept a second optional parameter `formatSpecifier` the nodes pass these specifiers down in the case of integer or double to string calls.
* adds two new nodes `String From Object And Format` and `String From Array And Format`.
* adds `DynamoPreferencesNumberFormat` as a special format that uses the preferences to set the specifier - maybe a simpler name for this `DYNFORMAT`? I am still on the fence about this.
* When we encounter a format exception - for example if you user tries to apply a binary format string to a floating point number (not allowed) - we catch the exception, log a warning on the node, and return the string "Format specifier was invalid." instead of the formatted string.
    * If this occurs to ANY data in an array, the entire array is not formatted.
* adds extended help docs for these two new nodes.
* new nodes are marked experimental.
* Introduces a new experimental state for nodes, if this is true a node will display an experimental glyph in the UI and there will be some extra text in the description.
*  Introduces a feature flag to control if the glyph should be displayed.


support all c# numeric specifiers:
![image](https://github.com/user-attachments/assets/1b18fcbb-e9fc-4576-a8a7-236fea7d170f)
a new experimental glyph is added  - we also display this glyph on nodes that are in the experimental category in the preferences - this is done with namespace matching.
![Screenshot 2024-11-06 194145](https://github.com/user-attachments/assets/ca63e3e5-4176-4ef3-bf8a-c39357cfbe9b)
we use the standard dotnet `experimental` attribute to control the experimental state - we also have a tooltip explaining what this means. Package authors can use this and it integrates into c# - if someone tries to use an experimental API it is a compile time error and they need to explicitly disable the warning to compile their c# code.
![Screenshot 2024-11-06 165924](https://github.com/user-attachments/assets/5ea445c1-12ff-49bf-b22f-6541b7dc9aa6)
also add a bit of text to extended description of the node in the library.
![Screenshot 2024-11-06 122100](https://github.com/user-attachments/assets/49987a82-d3fe-40de-8a26-bfd9dec8abf5)
extended docs for the new string format nodes are added and link to the super detailed c# docs.
![Screenshot 2024-11-06 121849](https://github.com/user-attachments/assets/15c290f3-311a-4fd0-9b7a-70afb37a2802)


TODO:
- [x] add tests
- [x] add more tests! 
- [x] cleanup method names
- [x] add localized resources
- [x] additional controls for using standard c# string formats -
- [x] assess with the teams if this should make it into 3.4 or is too risky
- [x] Currently the default format specifier is `f6` which keeps the behavior consistent with the old nodes... though I think actually `G` is a better default. It's what most users would expect IMO
- [ ] file a follow-up task for using format specifiers on reference types (classes)
- [ ] file a follow-up task for adding info or warnings if experimental methods are used in DS code or python.
- [ ] file a follow-up task for generating experimental docs urls.
- [x] tsplines - remove from experimental check




### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes

* add `String From` nodes that allows control over format of generated string using format specifiers.
* standard c# numeric specifiers are allowed. https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings
* in addition a custom Dynamo specifier is allowed that uses the user's current number format preferences in the dynamo preference. Using this is discouraged. `DynamoPreferencesNumberFormat`. (this is currently undocumented) I'd like insight from team members and ux if it makes sense.
* Dynamo now also supports the `system.diagnostics.codeanalysis.experimental` attribute https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.experimentalattribute?view=net-8.0 and will display a small experimental symbol on experimental nodes as well as adding some information to the description for these nodes.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

sm6srw reviewed (approved):
Looks good to me...

sm6srw commented:
But I missed that fact that you have failing tests....

mjkkirschner commented:
okay, so I figured out that the test failures are because my workaround for not creating a new node or migration doesn't work for deserializing 1.0 graphs in xml format. I don't really want to mess with that at this point. I will try introducing a new set of nodes and obsoleting the old ones instead without removing them.

pinzart90 commented on src/Engine/ProtoCore/DSASM/Mirror/ExecutionMirror.cs:
```diff
@@ -44,24 +45,91 @@ public ExecutionMirror(ProtoCore.DSASM.Executive exec, ProtoCore.RuntimeCore cor
         public string GetStringValue(StackValue val, Heap heap, int langblock, int maxArraySize, int maxOutputDepth, bool forPrint = false)
         {
-            if (formatParams == null)
-                formatParams = new OutputFormatParameters(maxArraySize, maxOutputDepth);
```
Is this member still useful/used anywhere?

pinzart90 commented on src/Engine/ProtoCore/Utils/StringUtils.cs:
```diff
@@ -1,12 +1,16 @@
 {
     public static class StringUtils
     {
+        internal const string DynamoPreferencesNumberFormat = nameof(DynamoPreferencesNumberFormat);
```
where is DynamoPreferencesNumberFormat defined?

mjkkirschner replied:
you are looking at it.
When this is the format specifier, we just use the current `ProtoCore.Mirror.MirrorData.PrecisionFormat` which is set by the prefs.

I actually think this is a bad idea but I believe it was part of the AC to control the output using the display number format.

I think it 's a bad idea because it makes sharing graphs harder.

pinzart90 replied:
Oh, I did not know you could define strings this way `symbol = nameof(symbol)`

Amoursol replied:
I'm fine with these as Experimental - Can we add in the details into an Extended Node Help @mjkkirschner ? 

Longer term agree this makes more sense with dropdowns and inputs that also work well with into ML Node Autocomplete.

mjkkirschner replied:
@Amoursol @Jingyi-Wen  - sure I can make sure these nodes have some .md files that explain how to use them.

As far as making sure users understand they are experimental - do you have any objections to naming them with `Experimental` in the name, or adding a new category for experimental nodes?

I think two big issues with the current experimental toggle we have are:
1. If I use the node from the library, I have no idea it's controlled by the experimental toggle.
2. Usually we make experimental features off by default so they have low discoverability where the whole point is to get as much feedback as possible.

Consider it an experiment to see how we can experiment better?

pinzart90 commented on src/Engine/ProtoCore/Lang/BuiltInMethods.cs:
```diff
@@ -130,6 +135,8 @@ public enum MethodID
             "__GC",                     // kGC
             Constants.kIfConditionalMethodName,
+            "__ToStringFromObjectAndFormat",     // kToStringFromObjectAndFormat
+            "__ToStringFromArrayAndFormat",      // kToStringFromArrayAndFormat
```
can't figure out why we need `__ToStringFromArrayAndFormat`. Can't replication work with `__ToStringFromObjectAndFormat` ? I must be missing something..

mjkkirschner replied:
yes replication works, but we need these overrides when you want a single string output, not a bunch of strings.

aparajit-pratap commented on src/Engine/ProtoCore/DSASM/Mirror/ExecutionMirror.cs:
```diff
@@ -44,24 +45,91 @@ public ExecutionMirror(ProtoCore.DSASM.Executive exec, ProtoCore.RuntimeCore cor
+                {
+                    return val.IntegerValue.ToString();
+                }
+                //TODO I am not sure if it's best to just return null or some message here
+                //or if we should catch the error higher up where we can log a runtime warning.
+                //currently we are doing the latter.
+
+                //return val.IntegerValue.SafeToStringWithFormat(formatSpecifier);
```
Can you explain this comment? Is it still relevant?

mjkkirschner replied:
yes, still an open question though I am leaning towards my current solution unless there are objections.
If we encounter an exception formatting, we throw and catch it the entry point of the built-in method, and set warning. I've seen this pattern for other built-ins.

aparajit-pratap commented on src/Engine/ProtoCore/DSASM/Mirror/ExecutionMirror.cs:
```diff
@@ -125,6 +193,8 @@ public string GetStringValue(StackValue val, Heap heap, int langblock, int maxAr
 
+        //TODO - use format specifier when converting to string
+        //obsolete this method and introduce another.
         public string GetClassTrace(StackValue val, Heap heap, int langblock, bool forPrint)
```
How about using the obsolete attribute?

aparajit-pratap commented on src/Engine/ProtoCore/DSASM/Mirror/ExecutionMirror.cs:
```diff
@@ -44,24 +45,91 @@ public ExecutionMirror(ProtoCore.DSASM.Executive exec, ProtoCore.RuntimeCore cor
+        /// <param name="forPrint"></param>
+        /// <param name="formatSpecifier"></param>
+        /// <returns></returns>
+        public string GetStringValueUsingFormat(StackValue val,string formatSpecifier, Heap heap, int langblock, bool forPrint = false)
```
Does this have to be public?

aparajit-pratap commented on src/Engine/ProtoCore/Lang/BuiltInMethods.cs:
```diff
@@ -797,6 +804,7 @@ public BuiltInMethods(Core core)
                         new KeyValuePair<string, ProtoCore.Type>("object", TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.Var, 0)),
                     },
                     ID = BuiltInMethods.MethodID.ToStringFromObject,
+                    MethodAttributes = new MethodAttributes(true, false, @"This method is obsolete, please use 'ToStringFromObjectAndFormat' "),
```
Can we make the existing `FromObject` nodemodel node point to your new builtin function instead (by passing in the old default for format specifier) and remove this builtin function altogether? Same question for the `FromArray` nodemodel node.

mjkkirschner replied:
oh.. hmm, I think I tried something slightly different, my attempt was to add ports to the existing nodes, though I guess we could try injecting the value and not messing with ports.

But - this will mean that existing ds code that calls `__StringFromObject` will no longer work. I don't know if that is a problem or not, but I thought it could be.

aparajit-pratap replied:
Do you mean in code block nodes or existing user-defined DS files? I don't remember if these functions with `__` can be called from CBNs.
I'm sure we'll need to update existing language tests but don't expect many cases of the builtin functions used directly by users.

mjkkirschner replied:
1. you can definitely call built-ins from codeblocks.
2. If this counts as a public function or not, I have no idea, `__` to me implies internal methods, but we've never communicated that to users - and because I want this PR to be as low risk as possible, I don't wish to remove them or mess with them, I want to introduce new nodes and mark them as experimental somehow to avoid accumulating more technical/API debt.
3. @Amoursol @jasonstratton - the sooner we can define an API boundary IMO the better to avoid these conversations and inconsistency for users.
![image](https://github.com/user-attachments/assets/fef01492-9a60-4c2f-a37d-3161558348e3)

aparajit-pratap replied:
Yes, they are and were meant to be internal functions. They were not meant to be publicly exposed to users, which is why they are wrapped as other nodes. Yes, they could have been exposed in code block nodes inadvertently, which may have been an oversight or carelessness on the Dynamo team but having said that IMO they can be safely removed to remove tech debt. I just wanted to let you know that I'm not asking you to remove them, that's up to you. I'm sorry if you don't like having these conversations but I will let folks know my opinion as long as I'm reviewing PRs.

mjkkirschner replied:
@aparajit-pratap - appreciate your opinion here on this particular issue of course - 

what I meant about "having these conversations" - was specifically that I want to get us to a place where we don't **need** to have a conversation each time we want to remove a method. It should be mechanical because something is either public and relied upon, or it's not. We intended to do this as a group after the 3.0 release - but have not yet done so formally.

I think it will be a boon to consistency for users and productivity for developers to get to this point.

aparajit-pratap commented on src/Engine/ProtoCore/Lang/BuiltInMethods.cs:
```diff
@@ -807,7 +815,35 @@ public BuiltInMethods(Core core)
+                        new KeyValuePair<string, ProtoCore.Type>("object", TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.Var, 0)),
+                        new KeyValuePair<string, ProtoCore.Type>("formatSpecifier", TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.String, 0)),
+                        //TODO (MJK)
+                        //if we wanted to support default args for builtins we would need to support these names being parsed as binary expressions instead
+                        //of just identifiers as is done today. We would need to add a new Parameters entry because all of this is public :(
+                        //or we could invoke the parser directly on these strings as is done for custom node symbols...
```
Can you explain this comment?

mjkkirschner replied:
```
        private static FunctionDefinitionNode GenerateBuiltInMethodSignatureNode(Lang.BuiltInMethods.BuiltInMethod method)
        {
            FunctionDefinitionNode fDef = new FunctionDefinitionNode();
            fDef.Name = Lang.BuiltInMethods.GetMethodName(method.ID);
            fDef.ReturnType = method.ReturnType;
            fDef.IsExternLib = true;
            fDef.IsBuiltIn = true;
            fDef.BuiltInMethodId = method.ID;
            fDef.Signature = new ArgumentSignatureNode();
            fDef.MethodAttributes = method.MethodAttributes;

            foreach (KeyValuePair<string, Type> param in method.Parameters)
            {
                VarDeclNode arg = new VarDeclNode();
                arg.NameNode = new IdentifierNode { Name = param.Key, Value = param.Key };
                arg.ArgumentType = param.Value;
                fDef.Signature.AddArgument(arg);
            }

            return fDef;
        }
        ```

These parameter strings are always just treated like identifiers, they don't get parsed into a full AST subtree where assignment to default values would normally be supported for function signatures.

aparajit-pratap commented on src/Libraries/CoreNodeModels/String.cs:
```diff
@@ -59,13 +62,50 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
         }
     }
 
+    [NodeName("String from Object And Format")]
```
How about renaming this to `Formatted String from Object`?

mjkkirschner replied:
I don't hate it!

aparajit-pratap commented on src/Libraries/CoreNodeModels/String.cs:
```diff
@@ -59,13 +62,50 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
     [OutPortTypes("string")]
     [IsDesignScriptCompatible]
     [AlsoKnownAs("DSCoreNodesUI.StringNodes.FromObject", "DSCoreNodesUI.FromObject")]
+    [Obsolete("this node is obsolete, please use the version of string from _ with numeric format option")]
```
?

mjkkirschner replied:
will clean these messages up, I actually think since I will mark the new nodes as experimental in some way I will leave this not obsolete.

aparajit-pratap commented on src/Libraries/CoreNodeModels/String.cs:
```diff
@@ -84,13 +123,51 @@ public FromObject() : base("__ToStringFromObject")
         }
     }
 
+    [NodeName("String from Array And Format")]
```
`Formatted String from Array`?

Amoursol commented:
@mjkkirschner @Jingyi-Wen color choice of the icon is a bit faded. Also green suggests good - when experimental should be neutral. Suggest switching to Blue here which is our "Informational" color. 

Also, styling of the hover tooltip is outdated @mjkkirschner

Amoursol commented:
@mjkkirschner could we also replace the Dynamo Icon in the Library with the Experimental one?

mjkkirschner commented:
> Also, styling of the hover tooltip is outdated @mjkkirschner
can you give me an example tooltip to checkout that has the correct styling?
> @mjkkirschner could we also replace the Dynamo Icon in the Library with the Experimental one?
I can look into this - if it's easy will do it otherwise will file an improvement.

I can change the color as long as I can find the info state color in the resource dictionary - @Jingyi-Wen any idea what else it's used for?

mjkkirschner commented:
> I can look into swapping the icon- if it's easy will do it otherwise will file an improvement.
actually, I don't think this such a great idea - we can control it from the feature flag, but I see two big downsides:
* Existing experimental nodes that have nice icons will now not show them in the library - for example the paneling nodes. Or when tsplines was experimental we would not have seen the icons - if we imagined that we did another tsplines like integration today and marked it experimental.
* If I am writing an experimental package or node I now don't have a way to test my icon easily - I guess locally I can swap the attribute out, but it's kind of a pain.

I will still evaluate how difficult it is to do technically, but it might benefit from ux thinking about it a bit more.

aparajit-pratap commented on src/Engine/ProtoCore/Utils/StringUtils.cs:
```diff
@@ -8,6 +8,7 @@ namespace ProtoCore.Utils
 {
     public static class StringUtils
     {
+        //TODO consider removing this option. It makes sharing harder.
```
This sort of looks similar to the scale factor setting issue where we have one in preferences and another setting in the workspace (graph), where the latter overrides the former. I would have suggested adding this format specifier at the graph level as well but I don't think that's needed if we're going to deprecate the old string nodes at some point.

aparajit-pratap commented on doc/distrib/NodeHelpFiles/CoreNodeModels.FormattedStringFromArray.md:
````diff
@@ -0,0 +1,17 @@
+Some commonly used format specifiers are:
+```
+G : general formatting G 1000.0 -> "1000"
+F : fixed point notation F4 1000.0 -> "1.0000"
````
Shouldn't this be "1000.0000"?

mjkkirschner replied:
yes, will fix, thanks.

aparajit-pratap commented on doc/distrib/NodeHelpFiles/CoreNodeModels.FormattedStringFromArray.md:
````diff
@@ -0,0 +1,17 @@
+```
+G : general formatting G 1000.0 -> "1000"
+F : fixed point notation F4 1000.0 -> "1.0000"
+N : number N 1.0 -> 1000 -> "1,000.00"
````
Did you mean this to be format specifier `N1`? If so, it should be 1000 -> 1,000.0

mjkkirschner replied:
so in an invariant culture `N` defaults to `N2` - but I should be clear since these are docs, I will update it.

Jingyi-Wen commented:
> > I can look into swapping the icon- if it's easy will do it otherwise will file an improvement.
> 
> actually, I don't think this such a great idea - we can control it from the feature flag, but I see two big downsides:
> 
> * Existing experimental nodes that have nice icons will now not show them in the library - for example the paneling nodes. Or when tsplines was experimental we would not have seen the icons - if we imagined that we did another tsplines like integration today and marked it experimental.
> * If I am writing an experimental package or node I now don't have a way to test my icon easily - I guess locally I can swap the attribute out, but it's kind of a pain.
> 
> I will still evaluate how difficult it is to do technically, but it might benefit from ux thinking about it a bit more. @Jingyi-Wen - maybe we only swap it out if the icon is not the default... or perhaps communicating that the node is experimental is more important than the icon - I'm not sure how to make that call.

@mjkkirschner I agree with this, experimental is more of a state instead of a symbol for the node itself. I'd suggest not swapping the icon in library
The blue color we use for info state is #38ABDF. Graph node manager has the color :)
Also, I suggest changing the tooltip wording to "This node is currently in the experimental state. Its behavior, name, and signature are subject to change."

aparajit-pratap commented on src/DynamoCore/PublicAPI.Unshipped.txt:
```diff
@@ -2782,6 +2782,7 @@ static Dynamo.Properties.Resources.DescriptionResource1.get -> string
 static Dynamo.Properties.Resources.DirectoryNotFound.get -> string
 static Dynamo.Properties.Resources.DisplayEngineFailureMessageDescription.get -> string
 static Dynamo.Properties.Resources.DllLoadException.get -> string
+static Dynamo.Properties.Resources.DocsExperimentalPrefixMessage.get -> string
```
Are we still seeing these code analysis warnings in PRs?

aparajit-pratap replied:
I expected to see them but lately, I've not been seeing them. I wonder if something changed. @avidit ?

mjkkirschner replied:
I saw them in the github build checks

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs:
```diff
@@ -641,7 +643,16 @@ public bool IsFrozenExplicitly
             }
         }
-
+        [Experimental("NVM_ISEXPERIMENTAL_GLPYH")]
```
Does the glyph need to be experimental?

mjkkirschner replied:
No, but it has to be public AFAK for wpf bindig and was not sure we'd want to support it. Experimental experimental mode ;)

aparajit-pratap reviewed (approved):
Minor comments and some questions. LGTM.

QilongTang commented on src/DynamoCore/Configuration/PreferenceSettings.cs:
```diff
@@ -112,6 +112,12 @@ private readonly static Lazy<PreferenceSettings>
+        internal static readonly IEnumerable<string> InitialExperimentalLib_Namespaces =
+        [
+            //TODO remove tsplines - it's no longer experimental.
+            "ProtoGeometry.dll:Autodesk.DesignScript.Geometry.TSpline",
```
Thanks, I thought I removed the TSpline feature flag check would be enough, I guess not..

QilongTang commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -145,7 +146,12 @@ public FunctionDescriptorParams()
         /// </summary>
-        public bool IsLacingDisabled { get; set; } 
+        public bool IsLacingDisabled { get; set; }
+        //TODO - should this somehow contain more info - ExperimentalInfo{IsExperimental, ExperimentalMessage/url}?}
```
How would the message be surfaced in your opinion?

mjkkirschner replied:
kind of like how the underlying `system.diagnostics.codeanalysis.experimental` attribute surfaces them. In this case it shows the errors in the errors tab of visual studio as compile time warnings and generates a url given a base url format and the experimental diagnostic id.

we could something similar with the docs browser, though it would be much more useful if you get a link to some real documentation that we update, for example in the forum or primer.

For now I will just leave it as a TODO and file a follow-up to consider it.

mjkkirschner commented on src/Libraries/CoreNodeModels/String.cs:
```diff
@@ -104,8 +107,6 @@ public FormattedStringFromObject() : base("__ToStringFromObjectAndFormat")
     [IsDesignScriptCompatible]
     [AlsoKnownAs("DSCoreNodesUI.StringNodes.FromObject", "DSCoreNodesUI.FromObject")]
-    [Obsolete("this node is obsolete, please use the version of string from _ with numeric format option")]
-    [NodeDeprecated]
```
My reasoning was as follows:

Since I am marking the new nodes experimental, I did not want to also mark the old nodes as obsolete (hiding them from the lib) - then it would not be clear to users how to proceed in a safe / future proof way.

I'd like to propose we follow this pattern when introducing new nodes (that replace existing ones):

phase 1: 
* Introduce new nodes experimental.
* Old nodes are untouched.
phase 2:
* gather feedback on experimental nodes.
* experimental nodes can be changed at any time.
phase 3:
* experimental attribute is removed from new nodes.
* old nodes are marked obsolete
phase 4:
* old nodes are removed in a major release.

mjkkirschner commented:
I think I have addressed all review comments, I will merge after I see a passing test result - at the moment a date time test is failing, I think @pinzart90 was trying to improve in another pr.

mjkkirschner commented:
passed here:
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16776/
